File: packages/custos-py/src/custos/telemetry.py
"""Opt-in anonymous telemetry (Python).

What is sent: {id, event, version, os, python} — nothing else.
Never sent: file paths, policy contents, tool names, ledger data, hostnames.

Off by default. Users must consent via `custos init`, or by editing
~/.custos/telemetry.json. Env var CUSTOS_TELEMETRY=off overrides consent.
No network requests unless CUSTOS_TELEMETRY_URL is set to a non-empty value.
"""
from __future__ import annotations

import json
import os
import platform
import sys
import threading
import uuid
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional
from urllib import request as urlrequest


CONFIG_DIR = Path.home() / ".custos"
CONFIG_PATH = CONFIG_DIR / "telemetry.json"


@dataclass
class TelemetryConfig:
    enabled: bool
    id: str
    consented_at: str
    version: int = 1


def _config_dir() -> Path:
    # Recomputed each call so tests that override HOME see the new location.
    return Path.home() / ".custos"


def _config_path() -> Path:
    return _config_dir() / "telemetry.json"


def read_config() -> Optional[TelemetryConfig]:
    p = _config_path()
    if not p.exists():
        return None
    try:
        data = json.loads(p.read_text(encoding="utf-8"))
        return TelemetryConfig(
            enabled=bool(data.get("enabled", False)),
            id=str(data.get("id", "")),
            consented_at=str(data.get("consented_at", "")),
            version=int(data.get("version", 1)),
        )
    except Exception:
        return None


def write_config(cfg: TelemetryConfig) -> None:
    d = _config_dir()
    d.mkdir(parents=True, exist_ok=True)
    _config_path().write_text(json.dumps(asdict(cfg), indent=2), encoding="utf-8")


def prompt_consent(assume_yes: bool = False, assume_no: bool = False) -> TelemetryConfig:
    existing = read_config()
    if existing is not None:
        return existing

    if assume_yes:
        enabled = True
    elif assume_no:
        enabled = False
    elif sys.stdin.isatty():
        sys.stdout.write(
            "Send anonymous usage counts to help prioritize Custos work? "
            "(only { install/command event, uuid, os, python version } — "
            "no paths, policies, or ledger data) [y/N] "
        )
        sys.stdout.flush()
        answer = sys.stdin.readline().strip().lower()
        enabled = answer in ("y", "yes")
    else:
        enabled = False

    cfg = TelemetryConfig(
        enabled=enabled,
        id=str(uuid.uuid4()) if enabled else "",
        consented_at=datetime.now(timezone.utc).isoformat(),
    )
    write_config(cfg)
    return cfg


def emit(event: str, cli_version: str) -> None:
    if os.environ.get("CUSTOS_TELEMETRY") == "off":
        return
    url = os.environ.get("CUSTOS_TELEMETRY_URL", "")
    if not url:
        return
    cfg = read_config()
    if cfg is None or not cfg.enabled or not cfg.id:
        return

    payload = json.dumps({
        "id": cfg.id,
        "event": event,
        "version": cli_version,
        "os": platform.system().lower(),
        "python": platform.python_version(),
    }).encode("utf-8")

    def send() -> None:
        try:
            req = urlrequest.Request(url, data=payload, headers={"content-type": "application/json"}, method="POST")
            urlrequest.urlopen(req, timeout=1.0).close()
        except Exception:
            pass

    threading.Thread(target=send, daemon=True).start()

File: packages/custos-js/src/telemetry.ts
// Opt-in anonymous telemetry.
//
// What is sent: { id, event, version, os, node } — nothing else.
// Never sent: file paths, policy contents, tool names, ledger data, hostnames.
//
// Off by default. Users must consent via `custos init` prompt, or by editing
// ~/.custos/telemetry.json. Env var CUSTOS_TELEMETRY=off overrides consent.
// No network requests unless CUSTOS_TELEMETRY_URL is set to a non-empty value.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { homedir, platform } from "node:os";
import { join } from "node:path";
import { randomUUID } from "node:crypto";
import { createInterface } from "node:readline";

const CONFIG_DIR = join(homedir(), ".custos");
const CONFIG_PATH = join(CONFIG_DIR, "telemetry.json");
const DEFAULT_URL = process.env.CUSTOS_TELEMETRY_URL ?? "";

export interface TelemetryConfig {
  enabled: boolean;
  id: string;
  consentedAt: string;
  version: 1;
}

export function readConfig(): TelemetryConfig | null {
  try {
    if (!existsSync(CONFIG_PATH)) return null;
    return JSON.parse(readFileSync(CONFIG_PATH, "utf8")) as TelemetryConfig;
  } catch {
    return null;
  }
}

export function writeConfig(cfg: TelemetryConfig): void {
  mkdirSync(CONFIG_DIR, { recursive: true });
  writeFileSync(CONFIG_PATH, JSON.stringify(cfg, null, 2));
}

export async function promptConsent(opts: { assumeYes?: boolean; assumeNo?: boolean } = {}): Promise<TelemetryConfig> {
  const existing = readConfig();
  if (existing) return existing;

  let enabled = false;
  if (opts.assumeYes) enabled = true;
  else if (opts.assumeNo) enabled = false;
  else if (process.stdin.isTTY) enabled = await ask();

  const cfg: TelemetryConfig = {
    enabled,
    id: enabled ? randomUUID() : "",
    consentedAt: new Date().toISOString(),
    version: 1,
  };
  writeConfig(cfg);
  return cfg;
}

async function ask(): Promise<boolean> {
  process.stdout.write(
    "Send anonymous usage counts to help prioritize Custos work? " +
    "(only { install/command event, uuid, os, node version } — no paths, policies, or ledger data) [y/N] ",
  );
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  return new Promise<boolean>((resolve) => {
    rl.question("", (answer) => {
      rl.close();
      resolve(/^y(es)?$/i.test(answer.trim()));
    });
  });
}

export interface EmitOptions {
  event: string;
  cliVersion: string;
}

export function emit(opts: EmitOptions): void {
  if (process.env.CUSTOS_TELEMETRY === "off") return;
  if (!DEFAULT_URL) return;
  const cfg = readConfig();
  if (!cfg || !cfg.enabled || !cfg.id) return;

  const payload = JSON.stringify({
    id: cfg.id,
    event: opts.event,
    version: opts.cliVersion,
    os: platform(),
    node: process.version,
  });

  // Fire-and-forget. Node < 18 lacks global fetch — bail silently there.
  const f = (globalThis as any).fetch as typeof fetch | undefined;
  if (!f) return;
  const ac = new AbortController();
  const t = setTimeout(() => ac.abort(), 1000);
  f(DEFAULT_URL, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: payload,
    signal: ac.signal,
  }).catch(() => {}).finally(() => clearTimeout(t));
}
